runtime/component/dnd-mode.js

```diff
@@ -0,0 +1,634 @@
+'use strict'
+
+var property = require('@yoda/property')
+var logger = require('logger')('dndmode')
+
+var FSMCode = {
+  Start: 0,
+  End: 254,
+  WaitAsync: 255,
+  CheckSwitchOn: 1,
+  CheckSwitchOff: 2,
+  CheckStatusOff: 3,
+  CheckStatusOn: 4,
+  DNDModeDisabled: 5,
+  CheckStatusOnX: 6,
+  CheckStatusOffX: 7,
+  CheckTimeSuccess: 8,
+  CheckTimeFailed: 9,
+  DNDModeDisabledX: 10,
+  CheckTimeSuccessX: 11,
+  CheckTimeFailedX: 12,
+  CheckActivityTrue: 13,
+  CheckActivityFalse: 14,
+  CheckActivityTrueX: 15,
+  CheckActivityFalseX: 16,
+  DNDModeEnabled: 17,
+  CheckAgain: 18
+}
+
+var SWITCH_KEY = 'dndmode.switch'
+var STATUS_KEY = 'dndmode.status'
+var WEEK_KEY = 'dndmode.weeksetting'
+var START_TIME_KEY = 'dndmode.starttime'
+var END_TIME_KEY = 'dndmode.endtime'
+var AWAKE_SWITCH_KEY = 'dndmode.awakeswitch'
+
+var FSM_READY = 0
+var FSM_RUNNING = 1
+var FSM_WAITING = 2
+var FSM_END = 3
+var FSM_ERROR = 4
+
+class DNDCommon {
+  constructor (life) {
+    this.life = life
+  }
+  /**
+   * Disable dnd mode
+   * @function disable
+   */
+  disable () {
+    // TODO disable dnd-mode
+    logger.info('dnd mode turned off')
+  }
+
+  /**
+   * Enable dnd mode
+   * @function enable
+   */
+  enable () {
+    // TODO enable dnd-mode
+    logger.info('dnd mode turned on')
+  }
+
+  /**
+   * Get dnd mode time span from now to start/end, return positive number if in dnd mode time
+   * @function getDNDTime
+   * @returns {number} - if result >= 0, it's the millisecond to end time
+   *                   - if result < 0, it's the millisecond to start time
+   */
+  static getDNDTime () {
+    function formatDate (dt) {
```

Comment:
This function could be declared outside of the class to reduce the operation of function creation.

Reply:
But this will cause more memory consumption. Is there any guiding principle in this regard?